media/multipart/common/src/test/java/io/helidon/media/multipart/common/MultiPartDecoderTest.java

```diff
@@ -0,0 +1,571 @@
+                + "body 1\n"
+                + "--" + boundary + "\n"
+                + "Content-Id: part2\n"
+                + "\n"
+                + "body 2\n"
+                + "--" + boundary + "--").getBytes();
+
+        final CountDownLatch latch = new CountDownLatch(4);
+        Consumer<BodyPart> consumer = (part) -> {
+            latch.countDown();
+            if (latch.getCount()== 3) {
+                assertThat(part.headers().values("Content-Id"), hasItems("part1"));
+                DataChunkSubscriber subscriber = new DataChunkSubscriber();
+                part.content().subscribe(subscriber);
+                subscriber.content().thenAccept(body -> {
+                    latch.countDown();
+                    assertThat(body, is(equalTo("body 1")));
+                });
+            } else {
+                assertThat(part.headers().values("Content-Id"), hasItems("part2"));
+                DataChunkSubscriber subscriber = new DataChunkSubscriber();
+                part.content().subscribe(subscriber);
+                subscriber.content().thenAccept(body -> {
+                    latch.countDown();
+                    assertThat(body, is(equalTo("body 2")));
+                });
+            }
+        };
+        BodyPartSubscriber testSubscriber = new BodyPartSubscriber(SUBSCRIBER_TYPE.ONE_BY_ONE, consumer);
+        partsPublisher(boundary, chunk1).subscribe(testSubscriber);
+        waitOnLatch(latch);
+        assertThat(testSubscriber.error, is(nullValue()));
+        assertThat(testSubscriber.complete, is(equalTo(true)));
+    }
+
+    @Test
+    public void testSubscriberCancelAfterOnePart() {
+        String boundary = "boundary";
+        final byte[] chunk1 = ("--" + boundary + "\n"
+                + "Content-Id: part1\n"
+                + "\n"
+                + "body 1\n"
+                + "--" + boundary + "\n"
+                + "Content-Id: part2\n"
+                + "\n"
+                + "body 2\n"
+                + "--" + boundary + "--").getBytes();
+
+        final CountDownLatch latch = new CountDownLatch(2);
+        Consumer<BodyPart> consumer = (part) -> {
+            latch.countDown();
+            if (latch.getCount()== 1) {
+                assertThat(part.headers().values("Content-Id"), hasItems("part1"));
+                DataChunkSubscriber subscriber1 = new DataChunkSubscriber();
+                part.content().subscribe(subscriber1);
+                subscriber1.content().thenAccept(body -> {
+                    latch.countDown();
+                    assertThat(body, is(equalTo("body 1")));
+                });
+            }
+        };
+
+        BodyPartSubscriber testSubscriber = new BodyPartSubscriber(SUBSCRIBER_TYPE.CANCEL_AFTER_ONE, consumer);
+        partsPublisher(boundary, chunk1).subscribe(testSubscriber);
+        waitOnLatch(latch);
+        assertThat(testSubscriber.error, is(nullValue()));
+        assertThat(testSubscriber.complete, is(equalTo(false)));
```

Comment:
Its not required by spec, is there business reason for this check?

Reply:
Probably not, I wasn't looking at making these spec compliant at the time. Thanks for pointing this out, I will remove these.

Reply:
This test is doing a cancel after receiving the first item ; the assertions test non of `onError()` and `onComplete()` were called.

Do you mean that it is legal for a subscriber to received either of these signals after the subscription has been canceled ?

The code that implements onError and onComplete checks for canceled subscription, that's the only reason for these 2 assertions.